### DISCUSSION TITLE
[Feedback Requested] Cross-platform Adaptation for Text Field

### DISCUSSION THREAD

**Discussion by InMatrix** [flutter/uxr#95] [Cross-platform Design]
## What is a text field?

Text fields, usually rectangular areas, allow users to enter text into a UI. They typically appear in forms and dialogs.

## Platform differences

Since Android 12, the default UI of text fields follows the design guidelines defined in [Material 3](https://m3.material.io/components/text-fields/overview) (M3). On iOS, an equivalent component is defined in [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/components/selection-and-input/text-fields/) (HIG).   There are some notable differences on the design of text fields between the two platforms.

### Text fields on Android 

On Android, Material Design text fields come in two basic styles: filled and outlined (see the screenshot below). They are more colorful and visually prominent than their iOS counterparts. 

<img src="https://github.com/flutter/uxr/assets/348942/3c24731d-ece1-4e25-a15e-ab32c3873d79" width=60% height=60%>

### Text fields on iOS

On iOS, text fields are unstyled by default, as you can see from the UI for entering recipients in the Messages app. 

<img src="https://github.com/flutter/uxr/assets/348942/9582129d-2aa6-476f-b04a-2beb446967df" width=60% height=60%>

It’s common to add thin gray borders to text fields in iOS apps. An example is the UI for entering message content in the Messages app. Rounded corners are used in the example, but they’re not inherently required by iOS. The examples in the text[ field’s HIG article](https://developer.apple.com/design/human-interface-guidelines/text-fields) do not have rounded corners. 

<img src="https://github.com/flutter/uxr/assets/348942/12b72c54-6b0c-4ee8-86b3-139ed5b9f386" width=60% height=60%>

The search field on iOS is a text field with a particular style. It has a filled background, rounded corners, but no border color. It also includes both leading and trailing icons in the text field. 

<img src="https://github.com/flutter/uxr/assets/348942/47336dd7-1c21-4822-a25f-5481b4d013e1" width=60% height=60%>

### Text fields in Flutter

				
The Flutter framework has two different widgets for displaying alert dialogs:

*   [TextField](https://api.flutter.dev/flutter/material/TextField-class.html), which implements the specification defined by Material Design.
*   [CupertinoTextField](https://api.flutter.dev/flutter/cupertino/CupertinoTextField-class.html), which follows the look & feel of the iOS default alert dialogs. 
## Adaptation recommendation

Our general recommendation is to follow the platform convention when using text fields in your app, because the default look of Material Design text fields has unusually prominent visual features and colors than most text fields on iOS. Switching to the `CupertinoTextField` is one way to do that, but you could also customize the `TextField` widget in the `material` library to achieve similar results with the benefits of sticking with the same API and having more customization options. Note that Flutter automatically adapts detailed interactions with text such as spell check and [text selection](https://docs.flutter.dev/resources/platform-adaptations#text-selection-toolbar) to the target platform for the `TextField` widget. 

The code snippets below show you how to do that, and you can see these two examples in action [in DartPad](https://dartpad.dev/?id=5e21d29ef2dc94ddad0a4a635679cfad). 

### Example code: basic text fields adaptive to iOS and Android

The code snippet below implements a basic text field showing a platform-specific UI through customizing `TextField`.

```dart
Widget _createAdaptiveTextField() {
  final _border = OutlineInputBorder(
    borderSide: BorderSide(color: CupertinoColors.lightBackgroundGray),
  );

  final iOSDecoration = InputDecoration(
    border: _border,
    enabledBorder: _border,
    focusedBorder: _border,
    filled: true,
    fillColor: CupertinoColors.white,
    hoverColor: CupertinoColors.white,
    contentPadding: EdgeInsets.fromLTRB(10, 0, 0, 0),
  );

  return Platform.isIOS
      ? SizedBox(
          height: 36.0,
          child: TextField(
            decoration: iOSDecoration,
          ),
        )
      : TextField();
}

```

### Example code: search fields adaptive to iOS and Android

The code snippet below implements a search field showing a platform-specific UI through customizing `TextField`.

```dart
Widget _createAdaptiveSearchField() {
  return Platform.isIOS? SizedBox(
    height: 36.0,
    child: TextField(
      autofocus: true,
      cursorColor: CupertinoColors.activeBlue,
      decoration: InputDecoration(
        prefixIcon: Icon(Icons.search),
        border: OutlineInputBorder(
          borderRadius: BorderRadius.circular(12.0),
          borderSide: BorderSide.none,
        ),
        hintText: 'Search',
        filled: true,
        fillColor: CupertinoColors.lightBackgroundGray,
        contentPadding: EdgeInsets.symmetric(vertical: 0.0),
        hintStyle: TextStyle(color: CupertinoColors.secondaryLabel),
        prefixIconColor: CupertinoColors.secondaryLabel,
      ),
    ),
  ): TextField(
    autofocus: true,
    decoration: InputDecoration(
      prefixIcon: Icon(Icons.search),
      border: OutlineInputBorder(),
      hintText: 'Search',
    ),
  );
}
```

Output on iOS: 

<img width="442" alt="a search field on iOS based on the TextField widget"  src="https://github.com/flutter/uxr/assets/348942/5619084e-63b3-4809-982b-f9f9b4ff8ea2">

Output on Android: 

<img width="445" alt="a search field on Android based on the TextField widget" src="https://github.com/flutter/uxr/assets/348942/ae73295f-0d24-4c67-914c-2b2cd070e2bc">

**Comment by laterdayi** [comment]
When will the Text Field adaptive api be added
  @InMatrix

**Reply by InMatrix** [reply]
> When will the Text Field adaptive api be added

@justinmc might be able to speak more to this question.

**Reply by justinmc** [reply]
There is no immediate timeline for this. Adaptive widgets is a big focus right now, but there is no PR on the way right now for this specifically. But I appreciate you chiming in to express your support and upvoting!

**Comment by laterdayi** [comment]
Looking forward to adding it, along with more adaptive apis, to add a better user experience
